feat: add samples for using Fluent Bit with Anthos Multi-Cloud

amandawestlake opened the pull request:
### Fixes ISSUE_NO_HERE
- N/A

#### Description
- Sample documents for a tutorial to use Fluent Bit to customize Cloud Logging logs for Anthos Multi-Cloud

#### Change summary
- Create sample files for Fluent Bit tutorial
- Edit anthos-multi-cloud README to add description of Fluent Bit files

#### Related PRs/Issues
- None

ckim328 commented:
Hi @amandawestlake ! Thanks for preparing this PR!
I'm seeing that the CI is failing on the checks for the licences, could you add them to the files you are adding? let me know if you have any questions :)

ckim328 commented on anthos-multi-cloud/AWS/vars.sh:
```diff
@@ -0,0 +1,21 @@
+export CLUSTER_NAME=aws-cluster-qa
```
Do we need to have these vars.sh files? I believe if anything, provide steps on how to get the appropriate variables in the readme?

amandawestlake replied:
Oh yes, those vars.sh files were added by accident!

amandawestlake replied:
I just removed them in the latest commit

ckim328 commented:
errors that im seeing in cloud build: 

> 
```
Checking for missing newline at end of file
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/kubernetes/test-logger.yaml
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/kubernetes/fluentbit-daemonset.yaml
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/kubernetes/fluentbit-rbac.yaml
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/kubernetes/fluentbit-configmap.yaml
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/README.md
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/test-logger/Dockerfile
Error: No newline at end of file ./anthos-multi-cloud/customize-logs-fluentbit/test-logger/go.mod
Error: No newline at end of file ./anthos-multi-cloud/README.md
```
>
```
Running shellcheck

In ./anthos-multi-cloud/AWS/vars.sh line 1:
export CLUSTER_NAME=aws-cluster-qa
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.
In ./anthos-multi-cloud/Azure/vars.sh line 1:
export CLUSTER_NAME=azure-cluster-gp
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.
```

Shabirmean commented on anthos-multi-cloud/README.md:
```diff
@@ -7,3 +7,7 @@ Checkout the different folders here for a Terraform based quickstart for setting
 - [Run GKE on AWS](/anthos-multi-cloud/AWS)
 - [Run GKE on Azure](/anthos-multi-cloud/Azure)
 
+The `customize-logs-fluentd` directory contains files to deploy a Fluent Bit
+DaemonSet on an Anthos Multi-Cloud cluster. Fluent Bit can be used to
+customize Cloud Logging logs for Anthos clusters on AWS and Anthos clusters
+on Azure.
```
Can we add this as a sub item to the list?

- [customize-logs-fluentd](/anthos-multi-cloud/customize-logs-fluentd): This directory contains files to deploy a Fluent Bit DaemonSet on an Anthos Multi-Cloud cluster. Fluent Bit can be used to customize Cloud Logging logs for Anthos clusters on AWS and Anthos clusters on Azure.

Shabirmean commented on anthos-multi-cloud/customize-logs-fluentbit/test-logger/logger.go:
```diff
@@ -0,0 +1,70 @@
+//Convert a string to an int, but consume any error and use the default instead.
+func convertToInt(s string, def int) int {
+	var result, err = strconv.Atoi(s)
+	fmt.Println(result)
```
This should go after the if check for `err`?

Shabirmean commented on anthos-multi-cloud/customize-logs-fluentbit/test-logger/logger.go:
```diff
@@ -0,0 +1,70 @@
+	return result
+}
+
+//Get a random log statement
```
Add a space after the comment directive `//Get` -> `// Get`

Shabirmean commented on anthos-multi-cloud/customize-logs-fluentbit/test-logger/logger.go:
```diff
@@ -0,0 +1,70 @@
+		logInfo(header)
+	}
+}
+func main() {
```
Add a space before the function main after the end pf the previous function

Shabirmean commented on anthos-multi-cloud/customize-logs-fluentbit/test-logger/logger.go:
```diff
@@ -0,0 +1,70 @@
+func main() {
+	rand.Seed(time.Now().UTC().UnixNano())
+	// Take the log interval and logging header from the environment
+	logInterval := convertToInt(os.Getenv("LOG_INTERVAL"), 2)
+	header := os.Getenv("HEADER")
```
Should we check and set default values if these were not set?

Shabirmean reviewed (commented):
Thank you @amandawestlake for opening this PR. I have left some comments. PTAL. I think we would want someone from the @GoogleCloudPlatform/anthos-multicloud  team to also have a review of the overall content.

CC: @bkauf @hdp617

amandawestlake commented:
Thank you @Shabirmean ! Just made commits addressing most of your comments. However, I did not write the .go file and I've also never used go before, so not sure if my changes are correct. Everything in the test-logger directory was copied directly from this repo: https://github.com/GoogleCloudPlatform/community/tree/master/tutorials/kubernetes-engine-customize-fluentbit.

Shabirmean reviewed (approved):
Thank you for the changes @amandawestlake 
The updates you made seems good to me :)

minherz commented on anthos-multi-cloud/customize-logs-fluentbit/kubernetes/fluentbit-configmap.yaml:
```diff
@@ -0,0 +1,171 @@
+        Match        k8s_application*
+        Condition    Key_value_equals stream stderr
+        Add          severity E
+  replace_info.lua: |
+        function replace_sensitive_info(tag, timestamp, record)
+          -- mask social security number
+          record["message"] = string.gsub(record["message"], "%d%d%d%-+%d%d%-+%d%d%d%d", "xxx-xx-xxxx")
+          -- mask credit card number
+          record["message"] = string.gsub(record["message"], "%d%d%d%d *%d%d%d%d *%d%d%d%d *%d%d%d%d", "xxxx xxxx xxxx xxxx")
+          -- mask email address
+          record["message"] = string.gsub(record["message"], "[%w+%.%-_]+@[%w+%.%-_]+%.%a%a+", "user@email.tld")
+          return 1, timestamp, record
+        end
```
nit: maybe this should be remarked as well, given that L121 comments script execution?

amandawestlake replied:
I'd prefer to leave this section uncommented so the user only has to uncomment one section, since there are already several steps involved with turning on log customization

minherz commented on anthos-multi-cloud/customize-logs-fluentbit/kubernetes/fluentbit-daemonset.yaml:
```diff
@@ -0,0 +1,115 @@
+        - --http_server=localhost:29681
+        - --ksa_name=user-telemetry-agent
+        - --ksa_namespace=logging-system
+        - --token_audience=${PROJECT_ID}.svc.id.goog
+        - --gcp_project_id=${PROJECT_ID}
+        - --gcp_sts_audience=identitynamespace:${PROJECT_ID}.svc.id.goog://gkemulticloud.googleapis.com/projects/${PROJECT_NUMBER}/locations/${REGION}/${CLUSTER_TYPE}/${CLUSTER_NAME}
```
nit: is it possible that environment variables used here aren't defined?

amandawestlake replied:
These samples are part of a tutorial that will be published on the community page that instructs users to set environment variables: https://github.com/amandawestlake/community/blob/fluentbit-anthosmulticloud/tutorials/customize-anthos-multicloud-logs-with-fluentbit/index.md
